### DISCUSSION TITLE
Enh: Add support for Model Context Protocol (MCP)

### DISCUSSION THREAD

**Discussion by flefevre** [open-webui/open-webui#7363] [General]
> [!TIP]
> # Documentation: https://docs.openwebui.com/openapi-servers/mcp
> # Open WebUI Integration: https://docs.openwebui.com/openapi-servers/open-webui
> # FAQ: https://docs.openwebui.com/openapi-servers/faq

**Comment by tjbck** [comment]
Supported via https://github.com/open-webui/openapi-servers#-bridge-mcp--openapi-optional with Open WebUI 0.6+

> [!TIP]
> Documentation: https://docs.openwebui.com/openapi-servers/mcp
> Open WebUI Integration: https://docs.openwebui.com/openapi-servers/open-webui

**Reply by flefevre** [reply]
Great job. I will try to test in one week!
Perhaps with the need for end users to be able to start a  filesystem access as executable as they are not python friendly in their window env.

**Reply by rgaricano** [reply]
why not? just intall python for w.
I left mwindows since ago, but I think that with a .bat it's more than enougth.
a exe it's a unknow cage/jail and command is comming back,
maybe users have to relearn to be friendly with open source.

**Reply by SecretiveShell** [reply]
Brilliant news!

**Reply by tamakiramimy** [reply]
When will this feature be available, and when will the OpenWeb 0.6+ version be released?

**Reply by dtinth** [reply]
I tested in latest Open WebUI v0.6.

The Open WebUI client-side makes 2 requests:

1. `GET /openapi.json`
2. `POST /<tool>`

It turns out that request #2 is missing an `Authorization` header, while request #1 has it. Not sure if others are facing the same problem but this is what I am encountering right now.

**Reply by Lutherwaves** [reply]
Thanks for adding this!

I think hardcoding to `/openapi.json` as the target endpoint should be changed to a setting specifying the endpoint where it is expected to keep things more flexible. In my experience and based on different usage patterns, different endpoints may be used (to comply with company practices etc).

I am testing this soon and will also pipe additional feedback.

**Reply by tjbck** [reply]
@dtinth @Lutherwaves addressed in dev!

**Reply by s4sanal85** [reply]
> @dtinth @Lutherwaves addressed in dev!

Thank you. Had seen the same problem

**Comment by rgaricano** [comment]
Testing and with some problem here and there,

I tried launching with mcpo, it was working on browser but open-webui doesn't see it, but  
it work when the tool is launched directly, running in tool directory: `uvicorn main:app --host 0.0.0.0 --port 8081 --reload`

(I tested with ssl/ssh too, and open-webui detect it fine: `uvicorn main:app --host 0.0.0.0 --port 8081 --reload --ssl-certfile="/usr/local/ssl/selfsigned.crt"  --ssl-keyfile="/usr/local/ssl/selfsigned.key"`)

(This it's the proccess with ssl/ssh (https:// ), I do in this way because I have all web entries proxied to 443 ssh and so I avoid problems, replace for http:// for a simple launch without certs)

launching tool, https in this case:

![Image](https://github.com/user-attachments/assets/7492cc55-71e3-4c94-9282-7e7503031242)

tool is exposed fine, i can access to swaggerUI,

![Image](https://github.com/user-attachments/assets/f452311b-169c-42c0-a03e-e878abd65aed)

& I see the openapi.json,

![Image](https://github.com/user-attachments/assets/745d1921-6c26-45fa-8272-87636c8f1cb7)

I set the connection on settings/tools (on admin, in user i can see server tools or option),
(first I tried whith ip 127.0.0.1, & i received a red flag, problem! )

![Image](https://github.com/user-attachments/assets/ff1a859f-0537-493c-a401-69a7bb3f520d)

As calls to tool servers are done by client side  https://github.com/open-webui/open-webui/issues/12313#issuecomment-2770605686 and  right now i didn't proxied any tool server, I made another try but now with the local IP of my server, it was working on browser 

![Image](https://github.com/user-attachments/assets/ee90d0fe-8ff9-43ad-8b8b-fa22fa1ebb6b)

It save it fine, without any red message!

![Image](https://github.com/user-attachments/assets/9818ddb2-e260-488a-a47b-ff47d08fe254)

Now I see the icon of mcp tool on chat window

![Image](https://github.com/user-attachments/assets/faf47ea5-d149-40ac-8d82-a360b0642e2c)

& I can see all function defined on server pressing on icon button

![MCP-Config-server7](https://github.com/user-attachments/assets/9f9f1a1a-d6fc-4102-b267-a400efad69bb)

& i receipt a valid response from model

![Image](https://github.com/user-attachments/assets/9d0b17ab-758c-4259-9c64-81d37566b16e)

**Reply by TheDiehard22** [reply]
I'd like to mention that some models need to have the "native" tool calling option enabled instead of default!
